java/org/apache/catalina/session/StandardSession.java

```diff
@@ -1548,25 +1588,68 @@ public boolean isAttributeDistributable(String name, Object value) {
     protected void doReadObject(ObjectInputStream stream)
         throws ClassNotFoundException, IOException {
 
+        // Test if we've got a CustomObjectInputStream; if so, disable filter
+        CustomObjectInputStream cois = null;
+        if (stream instanceof CustomObjectInputStream) {
+            cois = (CustomObjectInputStream) stream;
+        }
+
         // Deserialize the scalar instance variables (except Manager)
-        authType = null;        // Transient only
+        authType = null;        // Transient only (may be set later)
         creationTime = ((Long) stream.readObject()).longValue();
         lastAccessedTime = ((Long) stream.readObject()).longValue();
         maxInactiveInterval = ((Integer) stream.readObject()).intValue();
         isNew = ((Boolean) stream.readObject()).booleanValue();
         isValid = ((Boolean) stream.readObject()).booleanValue();
         thisAccessedTime = ((Long) stream.readObject()).longValue();
-        principal = null;        // Transient only
+        principal = null;        // Transient only (may be set later)
         //        setId((String) stream.readObject());
         id = (String) stream.readObject();
         if (manager.getContext().getLogger().isDebugEnabled())
             manager.getContext().getLogger().debug
                 ("readObject() loading session " + id);
 
+        // The next object read could either be the number of attributes (Integer) or a
```

Comment:
I agree with @markt-asf's suggestion that you just always persist the principal and authType for simplicity. Yes, it will not be backward-compatible, but how many servers do you have where you need to shut down down, upgrade then, and bring them back up (quickly!) and retain all of the existing sessions? It seems like more trouble than it's worth, and the code does not need to be this complicated.